test/services/blockchain/contracts/VestingContract.integration.test.ts

```diff
@@ -0,0 +1,250 @@
+                                .assignTokens('coinsale.tmy', accountName, '10.000000 LEOS', 1, signer)
+                                .then(resolve)
+                                .catch(reject);
+                        }, i * 2000); // Delay each transaction by 2 seconds
+                    })
+                );
+            }
+
+            await Promise.all(promises);
+        } catch (e) {
+            expect(e.message).toContain('Cannot purchase tokens more than 20 times.');
+        }
+    }, 60000);
+
+    test('withdraw(): Unsuccessful withdrawal vesting period after cliff not started', async () => {
+        expect.assertions(1);
+        const salesDate = '2023-03-27T00:00:00';
+        const launchDate = '2023-06-25T00:00:00';
+
+        try {
+            const { user } = await createRandomID();
+            const userAccountName = await user.storage.accountName;
+            const accountInfo = await getAccountInfo(userAccountName);
+            const accountName = accountInfo.account_name.toString();
+
+            await vestngContract.updatedate(salesDate, launchDate, signer);
+
+            await vestngContract.assignTokens('coinsale.tmy', accountName, '100.000000 LEOS', 1, signer);
+
+            await vestngContract.withdraw(accountName, signer);
+        } catch (e) {
+            expect(e.error.details[0].message).toContain('Vesting period after cliff has not started');
+        }
+    });
+
+    test('withdraw(): Successful withdrawal with cliff period', async () => {
+        expect.assertions(1);
+        const salesDate = '2023-03-20T00:00:00';
+        const launchDate = '2022-01-25T00:00:00';
+
+        const { user } = await createRandomID();
+        const userAccountName = await user.storage.accountName;
+        const accountInfo = await getAccountInfo(userAccountName);
+        const accountName = accountInfo.account_name.toString();
+
+        await vestngContract.updatedate(salesDate, launchDate, signer);
+        await vestngContract.assignTokens('coinsale.tmy', accountName, '100.000000 LEOS', 1, signer);
+
+        const trx = await vestngContract.withdraw(accountName, signer);
+
+        expect(trx.processed.receipt.status).toBe('executed');
+    });
+
+    test('withdraw(): Successful withdrawal with 2 different allocations of same category', async () => {
+        expect.assertions(2);
+
+        const category = 1;
+
+        for (let i = 0; i < 2; i++) {
+            const { user } = await createRandomID();
+            const userAccountName = await user.storage.accountName;
+            const accountInfo = await getAccountInfo(userAccountName);
+            const accountName = accountInfo.account_name.toString();
+
+            await vestngContract.assignTokens('eosio.token', accountName, '500.000000 LEOS', category, signer);
+            const trx = await vestngContract.withdraw(accountName, signer);
+
+            expect(trx.processed.receipt.status).toBe('executed');
```

Comment:
missing authority error for the withdrawal function as we have require_auth(holder) in withdraw